ledger/stateproofverificationtracker.go

```diff
@@ -0,0 +1,279 @@
+var (
+	errStateProofVerificationDataNotFound = errors.New("requested state proof verification data not found")
+)
+
+type verificationDeleteData struct {
+	confirmedRound      basics.Round
+	stateProofNextRound basics.Round
+}
+
+type verificationCommitData struct {
+	confirmedRound   basics.Round
+	verificationData ledgercore.StateProofVerificationData
+}
+
+// stateProofVerificationTracker is in charge of tracking data required to verify state proofs until such a time
+// as the data is no longer needed.
+type stateProofVerificationTracker struct {
+	// dbQueries are the pre-generated queries used to query the database, if needed,
+	// to lookup state proof verification data.
+	dbQueries stateProofVerificationDbQueries
+
+	// trackedCommitData represents the part of the tracked verification data currently in memory. Each element in this
+	// array contains both the data required to verify a single state proof and data to decide whether it's possible to
+	// commit the verification data to the database.
+	trackedCommitData []verificationCommitData
+
+	// trackedDeleteData represents the data required to delete committed state proof verification data from the
+	// database.
+	trackedDeleteData []verificationDeleteData
+
+	// stateProofVerificationMu protects trackedCommitData and trackedDeleteData.
+	stateProofVerificationMu deadlock.RWMutex
+
+	// log copied from ledger
+	log logging.Logger
+}
+
+func (spt *stateProofVerificationTracker) loadFromDisk(l ledgerForTracker, _ basics.Round) error {
+	preparedDbQueries, err := stateProofVerificationInitDbQueries(l.trackerDB().Rdb.Handle)
+	if err != nil {
+		return err
+	}
+
+	spt.dbQueries = *preparedDbQueries
+
+	spt.log = l.trackerLog()
+
+	spt.stateProofVerificationMu.Lock()
+	defer spt.stateProofVerificationMu.Unlock()
+
+	const initialDataArraySize = 10
+	spt.trackedCommitData = make([]verificationCommitData, 0, initialDataArraySize)
+	spt.trackedDeleteData = make([]verificationDeleteData, 0, initialDataArraySize)
+
+	return nil
+}
+
+func (spt *stateProofVerificationTracker) newBlock(blk bookkeeping.Block, delta ledgercore.StateDelta) {
+	currentStateProofInterval := basics.Round(blk.ConsensusProtocol().StateProofInterval)
+
+	if currentStateProofInterval == 0 {
+		return
+	}
+
+	if blk.Round()%currentStateProofInterval == 0 {
+		spt.insertCommitData(&blk)
+	}
+
+	if delta.StateProofNext != 0 {
+		spt.insertDeleteData(&blk, &delta)
+	}
+}
+
+func (spt *stateProofVerificationTracker) committedUpTo(round basics.Round) (minRound, lookback basics.Round) {
+	return round, 0
+}
+
+func (spt *stateProofVerificationTracker) produceCommittingTask(_ basics.Round, _ basics.Round, dcr *deferredCommitRange) *deferredCommitRange {
+	return dcr
+}
+
+func (spt *stateProofVerificationTracker) prepareCommit(dcc *deferredCommitContext) error {
+	spt.stateProofVerificationMu.RLock()
+	defer spt.stateProofVerificationMu.RUnlock()
+
+	lastDataToCommitIndex := spt.committedRoundToLatestCommitDataIndex(dcc.newBase)
```

Comment:
committedRoundToLatestCommitDataIndex method is only used here and its return value `lastDataToCommitIndex` is always used as `lastDataToCommitIndex+1`. Consider changing the method semantic and returning a value+1